sdk/src/impl/whirlpool-impl.ts

```diff
@@ -463,25 +489,33 @@ export class WhirlpoolImpl implements Whirlpool {
       whirlpool,
       tickLower,
       tickUpper,
+      tokenExtensionCtx,
     });
 
     const rewardsQuote = collectRewardsQuote({
       position: positionData,
       whirlpool,
       tickLower,
       tickUpper,
+      tokenExtensionCtx,
     });
 
     const shouldCollectFees = feesQuote.feeOwedA.gtn(0) || feesQuote.feeOwedB.gtn(0);
     invariant(
-      this.data.rewardInfos.length === rewardsQuote.length,
+      this.data.rewardInfos.length === rewardsQuote.rewardOwed.length,
       "Rewards quote does not match reward infos length"
     );
 
     const shouldDecreaseLiquidity = positionData.liquidity.gtn(0);
 
     const rewardsToCollect = this.data.rewardInfos
-      .filter((_, i) => (rewardsQuote[i] ?? ZERO).gtn(0))
+      .filter((_, i) => {
+        return (
+          (rewardsQuote.rewardOwed[i] ?? ZERO).gtn(0) ||
+          // we need to collect reward even if all reward will be deducted as transfer fee
```

Comment:
(don't have to change for this PR), but I think it might be good to add additional handling on the caller side to basically have a confirmation or force flag for reward and fee collections since there may be scenarios where it might benefit a user more to have a better understanding of how the fees may be calculated. 

One specific example, such as collecting protocol fees for our fee conversions.
If we have a script that regularly checks to see if there are uncollected fee conversions and attempts to collect too frequently, we might end up with less than optimal collection
Let's say that the transfer fee fee rate is `100%` (although this applies to lower percentages as well). 
We may end up in a state where every time we attempt to collect fees, we collect fees less than the `maximum_fee`, and end up actually receiving 0. vs. If we can detect on the caller that we would receive 0 fees, then we could try to optimize when to collect (letting the fees accrue beyond the `maximum_fee` to some threshold)

Reply:
yeah, good point. frequent collections can be a disadvantage in those cases.
(no update on this PR)